### ISSUE TITLE
Application freezes when a plugin is modifying many songs

### ISSUE THREAD

**Issue by CreamyCookie** [quodlibet/quodlibet#1851]
Quod Libet often freezes for several minutes when using a plugin on a lot of songs.

For example, just now the Force Write plugin took 5 minutes for 128 songs, most of the time being frozen. (i.e. QL is grayed out and unresponsive.) This also happens on other plugins, like the ReplayGain plugin, whenever a lot of files are affected, but only after the `WaitLoadWindow` with the text "Saving the songs you changed" appears.

My library is on an external, NTFS-formatted hard disk (on which I can write and read without a problem during the freeze), and I'm running QL from master, but it also happened in stable releases. I'm on Ubuntu 15.10 / 64bit.

**Comment by urielz** [comment]
That's weird. I haven't had any problems with this and I'm using the Force Write plugin frequently (as per your suggestion :)). It might be related to the external HDD? I just tried it on 2000 songs, it took 4 seconds on a solid state internal drive...

**Comment by lazka** [comment]
We've had several reports about slow write speed in mutagen regarding mmap + external HDD / NTFS / NFS. But I haven't figured out what's the problem there. If you use the newest mutagen (1.31, see the QL about dialog) things should be better as it has improved padding support resulting in less writes (especially for ogg/wma files)

**Comment by CreamyCookie** [comment]
>  I just tried it on 2000 songs, it took 4 seconds on a solid state internal drive...

Wow, nice! Yeah, it's true that this might only happen when the HDD is being slow, but I still think, that at least the freeze shouldn't happen. However my HDD shouldn't be too slow though, since it uses USB 3 and according to some testes can write about 100 MB/s. (Of course that's nothing against an SSD.)

> We've had several reports about slow write speed in mutagen regarding mmap + external HDD / NTFS / NFS. But I haven't figured out what's the problem there. If you use the newest mutagen (1.31, see the QL about dialog) things should be better as it has improved padding support resulting in less writes (especially for ogg/wma files)

I'm running master, so mutagen is 1.31, but it's interesting that the performance is so bad with NTFS. Is [this pull request](https://bitbucket.org/lazka/mutagen/pull-requests/11/util-use-mmapmap_private-when-resizing/diff) (about NTFS performance improvements) part of 1.31? Maybe I should switch to ext4, since I barely use Windows anyway.

**Comment by lazka** [comment]
> Wow, nice! Yeah, it's true that this might only happen when the HDD is being slow, but I still think, that at least the freeze shouldn't happen.

Yeah, we should do IO in threads (also when loading songs), but usually it isn't \_such\_ a big problem.

> I'm running master, so mutagen is 1.31, but it's interesting that the performance is so bad with NTFS. Is this pull request (about NTFS performance improvements) part of 1.31?

This turned out to be incorrect and got reverted

> Maybe I should switch to ext4, since I barely use Windows anyway. 

might just be the "external"/usb part... no guarantees that this will fix it ;)

**Comment by CreamyCookie** [comment]
> Yeah, we should do IO in threads (also when loading songs), but usually it isn't _such_ a big problem.

That's true. I guess, it wouldn't be an easy task though.

> might just be the "external"/usb part... no guarantees that this will fix it ;)

Good point. I wonder though, if this problem is solely caused by the external HDD. Shouldn't QL, even without threading, update the number of processed songs? For example, just now it stayed for more than a minute on 444/741 songs. The weird thing is that the first 400 songs or so were processed relatively fast.

**Comment by lazka** [comment]
> There's one weird thing regarding the freezing: Just now it stayed for more than a minute on "444/741 songs", but the first 400 songs or so were processed in a few seconds.

That could bet buffering in the kernel or fuse or filesystem

**Comment by declension** [comment]
My take-away on this: yes we should probably be doing threaded I/O, but we could (also?) use the QL tasks infrastructure when running plugins like Force Write on many songs (but also: force write should be default :) )

Not sure how, though. Some kind of hook in the base class for `SongsMenuPlugin`?

**Comment by CreamyCookie** [comment]
I'm closing this one since I didn't really have an issue with this for quite a while. (Not sure if I should create a new issue for threaded IO.)
